cirq/linalg/decompositions.py

```diff
@@ -854,3 +857,98 @@ def kak_decomposition(unitary_object: Union[np.ndarray, 'cirq.SupportsUnitary'],
         global_phase=g * inner_cannon.global_phase,
         single_qubit_operations_before=(b1, b0),
         single_qubit_operations_after=(a1, a0))
+
+
+def kak_vector(unitary: np.ndarray):
+    r"""Compute the KAK vectors of a two qubit unitary.
+
+    Any 2 qubit unitary may be expressed as
+
+    $$ U = k_l A k_r $$
+    where $k_l, k_r$ are single qubit (local) unitaries and
+    $$ A = \exp( i * \sum_{j=x,y,z} k_j \sigma_{j,0}\sigma{j,1} ) $$
+
+    where $(k_x,k_y,k_z)$ is in the Weyl chamber, a tetrahedron with corners
+    $$(0,0,0), (\pi/2,0,0), (\pi/4,pi/4,0), (\pi/4,\pi/4,\pi/4)$$
+
+    The vector entries are ordered such that
+    $$ \pi/2 - k_y \geq k_x \geq k_y \geq z \geq 0 $$
+
+    Args:
+        unitary: Shape (...,4,4) unitary numpy.ndarray.
+
+    Returns:
+        The KAK vector of the unitary, output shape (...,3).
+
+    References:
+        The appendix section of "Lower bounds on the complexity of simulating
+        quantum gates".
+        http://arxiv.org/abs/quant-ph/0307190v1
+    """
+    unitary = np.asarray(unitary)
+
+    UB = np.einsum('...ab,...bc,...cd', MAGIC_CONJ_T, unitary, MAGIC)
+
+    m = np.einsum('...ab,...cb', UB, UB)
+
+    evals, _ = np.linalg.eig(m)
+
+    # The algorithm in the appendix mentioned above is slightly incorrect in
+    # that it only works for elements of SU(4). A phase correction must be
+    # added to deal with U(4).
+    phases = np.log(-1j * np.linalg.det(unitary)).imag + np.pi / 2
```

Comment:
Could you benchmark the speed of this method applied to 1000 random unitaries (from `cirq.testing.random_unitary(4)`) and compare it to the speed of the existing decomposition method? So basically:

```
import time
unitaries = [cirq.testing.random_unitary(4) for _ in range(1000)]
t0 = time.monotonic()
_ = kak_vector(unitaries)
t1 = time.monotonic()
print(t1 - t0, 'secs')
```

Reply:
Nevermind this. I did a cursory check and got this:

New method: `0.15s` to do 10K instances.
Old method: `5.2s` to do 10K instances.
So basically a 30x speedup.

```
    import time
    unitaries = [cirq.testing.random_unitary(4) for _ in range(10000)]
    t0 = time.monotonic()
    _ = kak_vector(unitaries)
    t1 = time.monotonic()
    print(t1 - t0, 'secs')
    t0 = time.monotonic()
    for u in unitaries:
        _ = cirq.kak_decomposition(u)
    t1 = time.monotonic()
    print(t1 - t0, 'secs')
```